docs/architecture/adr-038-state-listening.md

```diff
@@ -20,31 +22,45 @@ In addition to these request/response queries, it would be beneficial to have a
 
 ## Decision
 
-We will modify the `MultiStore` interface and its concrete (`rootmulti` and `cachemulti`) implementations and introduce a new `listenkv.Store` to allow listening to state changes in underlying KVStores.
+We will modify the `CommitMultiStore` interface and its concrete (`rootmulti`) implementations and introduce a new `listenkv.Store` to allow listening to state changes in underlying KVStores.
 We will introduce a plugin system for configuring and running streaming services that write these state changes and their surrounding ABCI message context to different destinations.
 
-### Listening interface
+### Listening
 
-In a new file, `store/types/listening.go`, we will create a `WriteListener` interface for streaming out state changes from a KVStore.
+In a new file, `store/types/listening.go`, we will create a `MemoryListener` struct for streaming out protobuf encoded KV pairs state changes from a KVStore.
+The `MemoryListener` will be used internally by the concrete `rootmulti` implementation to collect state changes from KVStores.
 
 ```go
-// WriteListener interface for streaming data out from a listenkv.Store
-type WriteListener interface {
-	// if value is nil then it was deleted
-	// storeKey indicates the source KVStore, to facilitate using the same WriteListener across separate KVStores
-	// delete bool indicates if it was a delete; true: delete, false: set
-	OnWrite(storeKey StoreKey, key []byte, value []byte, delete bool) error
+// MemoryListener listens to the state writes and accumulate the records in memory.
+type MemoryListener struct {
+	stateCache []StoreKVPair
 }
-```
 
-### Listener type
+// NewMemoryListener creates a listener that accumulate the state writes in memory.
+func NewMemoryListener() *MemoryListener {
+	return &MemoryListener{}
+}
 
-We will create a concrete implementation of the `WriteListener` interface in `store/types/listening.go`, that writes out protobuf
-encoded KV pairs to an underlying `io.Writer`.
+// OnWrite implements MemoryListener interface
```

Comment:
Is this comment correct? I don't see a `MemoryListener` interface in this version of the ADR.

Reply:
Fixed.